### PR TITLE
fix event loop is closed

### LINKED ISSUE #5021: RuntimeError: Event loop is closed
### Steps to Reproduce

```python
import logging

from telegram.ext import (
    Application,
)

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)

logger = logging.getLogger(__file__)
logger.setLevel(logging.INFO)
logging.getLogger("httpx").setLevel(logging.WARNING)


def error_handler(update, context) -> None:
    logger.exception(
        msg="Unhanlded Exception while handling an update:", exc_info=context.error
    )


async def post_init(application: Application) -> None:
    await application.bot.send_message(OWNER_USER_ID, "deployed")


def main():
    application = (
        Application.builder()
        .token(TOKEN)
        .post_init(post_init)
        .read_timeout(20)
        .write_timeout(20)
        .connect_timeout(20)
        .pool_timeout(10)
        .get_updates_read_timeout(20)
        .build()
    )

    application.run_polling(
        allowed_updates=["message", "callback_query"], bootstrap_retries=10
    )


if __name__ == "__main__":
    main()
```

### Expected behaviour

The code should run normally.

### Actual behaviour

sometimes in unstable networks, i got `RuntimeError: Event loop is closed` error when running the code.

### Operating System

windows 11 24h2 with the code running in docker with python:3.13-slim image

### Version of Python, python-telegram-bot & dependencies

```shell
python-telegram-bot 22.5
Bot API 9.2
Python 3.13.9 (main, Oct 21 2025, 02:09:29) [GCC 14.2.0]
```

### Relevant log output

```python
/usr/local/lib/python3.13/site-packages/telegram/ext/_application.py:631: RuntimeWarning: coroutine 'Application._update_fetcher' was never awaited
  self._update_fetcher(), name=f"Application:{self.bot.id}:update_fetcher"
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/app/project/__main__.py", line 4, in <module>
    main()
    ~~~~^^
  File "/app/project/bot/bot_getupdates.py", line 79, in main
    application.run_polling(allowed_updates=["message", "callback_query"], bootstrap_retries=10)
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/telegram/ext/_application.py", line 839, in run_polling
    return self.__run(
           ~~~~~~~~~~^
        updater_coroutine=self.updater.start_polling(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<9 lines>...
        close_loop=close_loop,
        ^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/local/lib/python3.13/site-packages/telegram/ext/_application.py", line 1061, in __run
    loop.run_until_complete(self.start())
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 725, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "/usr/local/lib/python3.13/site-packages/telegram/ext/_application.py", line 631, in start
    self._update_fetcher(), name=f"Application:{self.bot.id}:update_fetcher"
                                                ^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/telegram/_bot.py", line 553, in id
    return self.bot.id
           ^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/telegram/_bot.py", line 542, in bot
    raise RuntimeError(
    ...<2 lines>...
    )
RuntimeError: ExtBot is not properly initialized. Call `ExtBot.initialize` before accessing this property.
Exception ignored in: <function Connection.__del__ at 0x7f954e1c2fc0>
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/site-packages/aiomysql/connection.py", line 1131, in __del__
  File "/usr/local/lib/python3.13/site-packages/aiomysql/connection.py", line 339, in close
  File "/usr/local/lib/python3.13/asyncio/selector_events.py", line 1205, in close
  File "/usr/local/lib/python3.13/asyncio/selector_events.py", line 865, in close
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 833, in call_soon
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 556, in _check_closed
RuntimeError: Event loop is close
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by codomposer**
## Fixes

Closes #5021 

## Check-list for PRs

- [ ] Added `.. versionadded:: NEXT.VERSION`, ``.. versionchanged:: NEXT.VERSION``, ``.. deprecated:: NEXT.VERSION`` or ``.. versionremoved:: NEXT.VERSION` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)
- [x] Created new or adapted existing unit tests
- [x] Documented code changes according to the [CSI standard](https://standards.mousepawmedia.com/en/stable/csi.html)
- [ ] Added myself alphabetically to `AUTHORS.rst` (optional)
- [ ] Added new classes & modules to the docs and all suitable ``__all__`` s
- [ ] Checked the [Stability Policy](https://docs.python-telegram-bot.org/stability_policy.html) in case of deprecations or changes to documented behavior

**If the PR contains API changes (otherwise, you can ignore this passage)** - N/A

- New Classes
    - [ ] Added `self._id_attrs` and corresponding documentation
    - [ ] [__init__](cci:1://file:///c:/Users/Administrator/Documents/bt/sn74-gittensor/python-telegram-bot/src/telegram/_bot.py:313:4-391:22) accepts `api_kwargs` as keyword-only

- Added New Shortcuts
    - [ ] In [`telegram.Chat`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.chat.html) \& [`telegram.User`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.user.html) for all methods that accept `chat/user_id` 
    - [ ] In [`telegram.Message`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.message.html) for all methods that accept `chat_id` and `message_id` 
    - [ ] For new `telegram.Message` shortcuts: Added `quote` argument if methods accept `reply_to_message_id` 
    - [ ] In [`telegram.CallbackQuery`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.callbackquery.html) for all methods that accept either `chat_id` and `message_id` or `inline_message_id` 

- If Relevant
    - [ ] Added new constants at `telegram.constants` and shortcuts to them as class variables
    - [ ] Linked new and existing constants in docstrings instead of hard-coded numbers and strings
    - [ ] Added new message types to `telegram.Message.effective_attachment` 
    - [ ] Added new handlers for new update types
        - [ ] Added the handlers to the warning loop in the [`telegram.ext.ConversationHandler`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.conversationhandler.html)
    - [ ] Added new filters for new message (sub)types
    - [ ] Added or updated documentation for the changed class(es) and/or method(s)
    - [ ] Added the new method(s) to `_extbot.py` 
    - [ ] Added or updated `bot_methods.rst` 
    - [ ] Updated the Bot API version number in all places: `README.rst` (including the badge) and `telegram.constants.BOT_API_VERSION_INFO` 
    - [ ] Added logic for arbitrary callback data in `telegram.ext.ExtBot` for new methods that either accept a `reply_markup` in some form or have a return type that is/contains [`telegram.Message`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.message.html)

Contribution by Gittensor, learn more at https://gittensor.io/

### CODE REVIEW

File: src/telegram/ext/_utils/networkloop.py [hunk 1]
```diff
@@ -141,13 +141,52 @@ async def do_action() -> None:
...
+            # Check max_retries for TimedOut as well
+            if on_err_cb:
+                on_err_cb(toe)
+
+            if 0 <= max_retries <= retries:
+                _LOGGER.exception(
+                    "%s Failed run number %s of %s. Aborting.", log_prefix, retries, max_retries
+                )
+                raise
+        except InvalidToken as invalid_token_exc:
```
Comment (Bibo-Joshi) [inline src/telegram/ext/_utils/networkloop.py]: For Invalid token it doesn't actually make sense to change anything. I was a bit too eager in my comment https://github.com/python-telegram-bot/python-telegram-bot/issues/5021#issuecomment-3466643711, there. If the token is invalid, we simply exit as before, as there's no sense in doing any retries

File: src/telegram/ext/_utils/networkloop.py [hunk 2]
```diff
@@ -141,13 +141,52 @@ async def do_action() -> None:
...
+
+            # Check max_retries for RetryAfter as well
+            if on_err_cb:
+                on_err_cb(exc)
+
+            if 0 <= max_retries <= retries:
+                _LOGGER.exception(
+                    "%s Failed run number %s of %s. Aborting.", log_prefix, retries, max_retries
+                )
+                raise
```
Comment (Bibo-Joshi) [inline src/telegram/ext/_utils/networkloop.py]: 1) we should also have a log issued in case we do retry - see how it's handled in the `TelegramError` clause
2) please avoid duplication of this `if-else` logic, as it's now repeated in `RetryAfter`, `TimedOut` and `TelegramError`. One option would be to intrudce a small local helper function that checks the max-try condition and issues the logs accordingly

File: src/telegram/ext/_utils/networkloop.py [hunk 3]
```diff
@@ -141,13 +141,52 @@ async def do_action() -> None:
             )
             # pylint: disable=protected-access
             cur_interval = slack_time + exc._retry_after.total_seconds()
+
+            # Check max_retries for RetryAfter as well
+            if on_err_cb:
+                on_err_cb(exc)
```
Comment (Bibo-Joshi) [inline src/telegram/ext/_utils/networkloop.py]: `on_err_cb` should be called for `TelegramError` s only, not for `RetryAfter`, `TimedOut` or `InvalidToken`.

File: src/telegram/ext/_utils/networkloop.py [hunk 4]
```diff
@@ -103,6 +103,25 @@ async def network_retry_loop(
     log_prefix = f"Network Retry Loop ({description}):"
     effective_is_running = is_running or (lambda: True)
 
+    def check_max_retries_and_log(current_retries: int) -> bool:
```
Comment (Bibo-Joshi) [inline src/telegram/ext/_utils/networkloop.py]: this looks good, but we'd now have duplicated logs - the logs in lines 113 & 121 are issued in addition to e.g. line 168. We could even have situations of logs:

```
Timed out: %s. Retrying immediately
Failed run number %s of %s. Aborting.
```

which is rather confusing :D Please extend this function to accept strings & exception objects as necessary to avoid the duplication. My goal would be:

1. the if-else logic in lines 112-122 stays as is
2. the two possible logs indicate the number of the current run & the number of max runs
3. the two possible logs indicate the additional information currently logged for RetryAfter an TimedOut

File: tests/test_bot.py [hunk 5]
```diff
@@ -4633,3 +4633,83 @@ async def test_get_my_star_balance(self, bot):
         balance = await bot.get_my_star_balance()
         assert isinstance(balance, StarAmount)
         assert balance.amount == 0
+
+    async def test_initialize_partial_failure_recovery(self, offline_bot):
```
Comment (Bibo-Joshi) [inline tests/test_bot.py]: please move these tests up, below `test_multiple_inits_and_shutdowns`

File: tests/test_bot.py [hunk 6]
```diff
@@ -4633,3 +4633,83 @@ async def test_get_my_star_balance(self, bot):
...
+        """Test that bot properly tracks partial initialization and can recover."""
+        test_bot = PytestBot(token=offline_bot.token, request=OfflineRequest())
+
+        try:
+            # Manually simulate partial initialization state
+            # Initialize requests but not bot user
+            await asyncio.gather(
+                test_bot._request[0].initialize(), test_bot._request[1].initialize()
```
Comment (Bibo-Joshi) [inline tests/test_bot.py]: Please avoid relying on protected attributes (anything with a leading underscore). We want to test the behavior, not the internal implementation : ) You can use `monkeypatch` to e.g. mock the init/shutdown calls of the request objects as you've already done in the tests below. The goal would be to simply assert that the httpxrequest.init/shutdown functions as well as get_me have been called as expected. Asserting the value of protected members is not necessary.

this also holds for the other tests

File: tests/ext/_utils/test_networkloop.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains tests for the network_retry_loop function."""
+
+import pytest
+
+from telegram.error import InvalidToken, RetryAfter, TelegramError, TimedOut
+from telegram.ext._utils.networkloop import network_retry_loop
+
+
+class TestNetworkRetryLoop:
```
Comment (Bibo-Joshi) [inline tests/ext/_utils/test_networkloop.py:33]: Please use `pytest.mark.parametrize` to reduce the code repetition in these tests.
please also levave some comments saying that most part of the retry-loops functionality are covered in test_updater and test_application

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Thanks for the updates. I've left some new comments below. Please let me know in case you have any questions on them. In addition

* please have a look at the test coverage report https://app.codecov.io/gh/python-telegram-bot/python-telegram-bot/pull/5030. It should also show in the CI checks on github usually. Don't know why it doesn't so far …
* please add a file `changes/unreleased/5030.bBRC3598kUYHL2Qpo4gjMb.toml` with the content

```toml
bugfixes = "Fix a Bug in Initialization Logic of ``Bot``"
[[pull_requests]]
uid = "5030"
author_uids = ["codomposer"]
closes_threads = ["5021"]
```

**Comment by codomposer** [comment]
@Bibo-Joshi 
can you please check this PR again?

**Comment by Bibo-Joshi** [comment]
@codomposer Hey. I will come back to this, when I have the time and mental capacity for it, yes.

File: tests/ext/_utils/test_networkloop.py [hunk 8]
```diff
@@ -16,7 +16,13 @@
 #
 # You should have received a copy of the GNU Lesser Public License
 # along with this program.  If not, see [http://www.gnu.org/licenses/].
-"""This module contains tests for the network_retry_loop function."""
+"""This module contains tests for the network_retry_loop function.
+
+Note:
+    Most of the retry loop functionality is already covered in test_updater and test_application.
+    These tests focus specifically on the max_retries behavior for different exception types
+    and the error callback handling, which were added as part of the bug fix.
```
Comment (Bibo-Joshi) [inline tests/ext/_utils/test_networkloop.py]:
```suggestion
    and the error callback handling, which were added as part of the bug fix in #5030
```

File: tests/test_bot.py [hunk 9]
```diff
@@ -373,6 +373,53 @@ async def shutdown(*args, **kwargs):
...
+            assert get_me_call_count == 1
+
+            # Second initialize attempt should succeed
+            await test_bot.initialize()
+            # Request initialization should not be called again (still 1)
+            assert request_init_count == 1
+            # get_me should have been called a second time and succeeded
+            assert get_me_call_count == 2
+            # Verify bot is now accessible
+            assert test_bot.bot.id == offline_bot.id
```
Comment (Bibo-Joshi) [inline tests/test_bot.py:419]: please add one more more call to `init()` and check the call counts. that will cover the case that `if not self._bot_initialized` is `False`. That branch is currently not covered according to codecov :)

https://app.codecov.io/gh/python-telegram-bot/python-telegram-bot/pull/5030#42bb73a1b54e6a774669799b44ac6ebb-R854

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Hi. I've just 2 small comments open. Thanks very much for the updates and sorry for the delay in review - busy week …

**Comment by codomposer** [comment]
> Hi. I've just 2 small comments open. Thanks very much for the updates and sorry for the delay in review - busy week …

comments addressed, can you please check it? @Bibo-Joshi

**Comment by Bibo-Joshi** [comment]
Hi. Unfortunately, the if-clause is still not correctly covered. Please use pytest-cov locally to compute the coverage and verify that the case is covered. If you need help with that, please reach out. This week, I'm unlikely to have to to dig into that myself …

**Comment by codomposer** [comment]
> Hi. Unfortunately, the if-clause is still not correctly covered. Please use pytest-cov locally to compute the coverage and verify that the case is covered. If you need help with that, please reach out. This week, I'm unlikely to have to to dig into that myself …

@Bibo-Joshi 
I've pushed update to cover test_initialize_with_bot_already_initialized_but_not_requests,
can you comment under codeblocks that are missing test coverage if any?
thanks

**Comment by Bibo-Joshi** [comment]
Ah, I see my mistake now: `get_me` can only be called if the request init already succeeded and only then can `_bot_initialized` become `True`. This means that we simply don't need the `if not self._bot_initialized` check. In the test this shows because you can only catch the current edge case by adapting the internals of Bot, not by mocking the methods called by `Bot.init`.
TL;DR: please revert your last commit and instead remove the mentioned if-clause. Please excuse the confusion.

**Comment by codomposer** [comment]
> Ah, I see my mistake now: `get_me` can only be called if the request init already succeeded and only then can `_bot_initialized` become `True`. This means that we simply don't need the `if not self._bot_initialized` check. In the test this shows because you can only catch the current edge case by adapting the internals of Bot, not by mocking the methods called by `Bot.init`. TL;DR: please revert your last commit and instead remove the mentioned if-clause. Please excuse the confusion.

done

**Comment by Bibo-Joshi** [comment]
Thank you very much for your contribution, and also for your patience in the review process! 🙌
